From 3fa277b479d69699bf5a6875cd4a5efcf9ae0788 Mon Sep 17 00:00:00 2001
From: Alexey Sheplyakov <asheplyakov@mirantis.com>
Date: Tue, 27 Jun 2017 16:07:01 +0400
Subject: [PATCH] jewel: osd: unlock sdata_op_ordering_lock with sdata_lock
 hold to avoid missing wakeup signal

Based on commit bc683385819146f3f6f096ceec97e1226a3cd237. The OSD code has
been refactored a lot since Jewel, hence cherry-picking that patch introduces
a lot of unrelated changes, and is much more difficult than reusing the idea.

Fixes: http://tracker.ceph.com/issues/20428

Signed-off-by: Alexey Sheplyakov <asheplyakov@mirantis.com>
---
 src/osd/OSD.cc | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/osd/OSD.cc b/src/osd/OSD.cc
index f5cfda3b686a..38a2711f6f92 100644
--- a/src/osd/OSD.cc
+++ b/src/osd/OSD.cc
@@ -8727,9 +8727,9 @@ void OSD::ShardedOpWQ::_process(uint32_t thread_index, heartbeat_handle_d *hb )
   assert(NULL != sdata);
   sdata->sdata_op_ordering_lock.Lock();
   if (sdata->pqueue->empty()) {
-    sdata->sdata_op_ordering_lock.Unlock();
     osd->cct->get_heartbeat_map()->reset_timeout(hb, 4, 0);
     sdata->sdata_lock.Lock();
+    sdata->sdata_op_ordering_lock.Unlock();
     sdata->sdata_cond.WaitInterval(osd->cct, sdata->sdata_lock, utime_t(2, 0));
     sdata->sdata_lock.Unlock();
     sdata->sdata_op_ordering_lock.Lock();
